### PR TITLE
'string["substring"]' (like in Ruby) added

### PR DESCRIPTION
**Pull request by epitron**
A nice convenient way of checking the presence of a substring in a string.

### CODE REVIEW

File: src/string.cr [hunk 1]
```diff
@@ -160,6 +160,16 @@ class String
...
   end
 
+  def [](str : String)
+    if start = index(str)
+      self[start...start+str.size]
+    end
+  end
+
+  def []?(str : String)
+    !!index(str)
```
Comment (jhass) [inline src/string.cr]: So far the convention for `#[]?` and `#[]` is that `#[]` returns the value or raises and `#[]?` returns the value or `nil`.

Comment (asterite) [inline src/string.cr]: Yes. Also, is the slice really needed? I would have thought the implementation would be something like this:

```crystal
def [](str : String)
  self[str]?.not_nil! # or raise a better error message
end

def []?(str : String)
  includes?(str) ? str : nil
end
```

**Comment by epitron** [comment]
As a discussion topic for later: shouldn't all the []? methods return bools?

**Comment by asterite** [comment]
The convention is for `obj[key]` to return a value or else raise if `key` is missing (the definition of "missing" depends on the type of `obj`) and for `obj[key]?` to return a value or else nil if `key` is missing.

For other methods, it depends. If there's a method named `foo` and another `foo?` for the same type, it means that `foo` will raise on some condition while `foo?` will return nil in that same condition. If there's just the `foo?` variant but no `foo`, it returns a truthy or falsey value (not necessarily `true` or `false`).

Examples for all of the above:

* `Array#[](index)` raises on out of bounds, `Array#[]?(index)` returns nil in that case.
* `Hash#[](key)` raises if the key is not in the hash, `Hash#[]?(key)` returns nil in that case.
* `Array#first` raises if the Array is empty (there's no "first", so "first" is missing), while `Array#first?` returns nil in that case. Same goes for pop/pop?, shift/shift?, last/last?
* There's `String#includes?(obj)`, `Enumerable#includes?(obj)` and `Enumerable#all?`, all of which don't have a non-question variant. The previous methods do indeed return true or false, but that is not a necessary condition.

The only inconsistency we have, I think, is with `Enumerable#index(obj)`, which returns nil if obj is not present. Maybe we should fix this?

By the way, I added the above explanation to [this wiki page](https://github.com/manastech/crystal/wiki/Crystal-for-Rubyists).

File: src/string.cr (hunk 1 above)
Comment (epitron) [inline src/string.cr]: Hahah.. that's true. I wasn't thinking clearly. I'll fix. :)

**Comment by epitron** [comment]
Ah, I see. Thanks for laying that all out. This makes sense in the context of a compiled language.

**Comment by epitron** [comment]
The pull request has been fixed! (Sorry, I didn't realize force-pushing would kill all the comments.)

**Comment by asterite** [comment]
Thanks!

**Pull request merged**